tests/components/caldav/test_config_flow.py

```diff
@@ -0,0 +1,238 @@
+"""Test the Caldav config flow."""
+
+from unittest.mock import patch
+
+from caldav.lib.error import DAVError
+
+from homeassistant.components.caldav.const import (
+    CONF_ADD_CUSTO_CALENDAR,
+    CONF_CALENDAR,
+    CONF_CALENDARS,
+    CONF_CUSTOM_CALENDARS,
+    CONF_DAYS,
+    CONF_SEARCH,
+    DOMAIN,
+)
+from homeassistant.config_entries import SOURCE_IMPORT, SOURCE_USER
+from homeassistant.const import (
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_SOURCE,
+    CONF_URL,
+    CONF_USERNAME,
+    CONF_VERIFY_SSL,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import (
+    RESULT_TYPE_ABORT,
+    RESULT_TYPE_CREATE_ENTRY,
+    RESULT_TYPE_FORM,
+)
+
+from tests.common import MockConfigEntry
+
+USER_INPUT = {
+    CONF_USERNAME: "username",
+    CONF_PASSWORD: "password",
+    CONF_URL: "http://homeassistant.io",
+    CONF_VERIFY_SSL: True,
+    CONF_DAYS: 2,
+}
+
+USER_INPUT_CUSTO = {
+    CONF_USERNAME: "username",
+    CONF_PASSWORD: "password",
+    CONF_URL: "http://homeassistant.io",
+    CONF_VERIFY_SSL: True,
+    CONF_DAYS: 2,
+    CONF_ADD_CUSTO_CALENDAR: True,
+}
+
+CUSTOM_INPUT = {
+    CONF_CUSTOM_CALENDARS: [
+        {CONF_NAME: "name", CONF_CALENDAR: "calendar", CONF_SEARCH: "search"}
+    ]
+}
+
+IMPORT_INPUT = {
+    CONF_USERNAME: "username",
+    CONF_PASSWORD: "password",
+    CONF_URL: "http://homeassistant.io",
+    CONF_VERIFY_SSL: True,
+    CONF_DAYS: 2,
+    CONF_CALENDARS: ["home", "assistant"],
+    CONF_CUSTOM_CALENDARS: [
+        {CONF_NAME: "name", CONF_CALENDAR: "calendar", CONF_SEARCH: "search"},
+        {CONF_NAME: "name_1", CONF_CALENDAR: "calendar_1", CONF_SEARCH: "search_1"},
+    ],
+}
+
+OPTIONS_INPUT = {
+    CONF_CALENDARS: ["home", "assistant"],
+    CONF_CUSTOM_CALENDARS: [
+        {CONF_NAME: "name", CONF_CALENDAR: "calendar", CONF_SEARCH: "search"},
+        {CONF_NAME: "name_1", CONF_CALENDAR: "calendar_1", CONF_SEARCH: "search_1"},
+    ],
+}
+
+
+async def test_user_form(hass):
+    """Test we get the user initiated form."""
+
+    with patch(
+        "homeassistant.components.caldav.config_flow.async_caldav_connect",
+        return_value=[],
+    ):
```

Comment:
Please patch for only parts that need the patch in place or move the patch to be a decorator.  You can `await hass.async_block_till_done()` inside the with block to make sure its the action is completed before outdenting